internal/pkg/tablemgr/redisDeltaSyncer.go

```diff
@@ -0,0 +1,297 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tablemgr
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-redis/redis"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+	"github.com/xmidt-org/ears/internal/pkg/logs"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+
+	// used by one instance of ears to ask others to sync routing table
+
+	EARS_REDIS_SYNC_CHANNEL = "ears_sync"
+
+	// used by one instance of ears to tell all others that it just finished syncing its routing table
+
+	EARS_REDIS_ACK_CHANNEL = "ears_ack"
+
+	EARS_REDIS_RETRY_INTERVAL_SECONDS = 10 * time.Second
+)
+
+type (
+	RedisDeltaSyncer struct {
+		sync.Mutex
+		redisEndpoint     string
+		active            bool
+		client            *redis.Client
+		localTableSyncers map[RoutingTableLocalSyncer]struct{}
+		logger            *zerolog.Logger
+		config            Config
+	}
+)
+
+func NewRedisDeltaSyncer(logger *zerolog.Logger, config Config) RoutingTableDeltaSyncer {
+	s := new(RedisDeltaSyncer)
+	s.logger = logger
+	s.config = config
+	s.redisEndpoint = config.GetString("ears.synchronization.endpoint")
+	s.localTableSyncers = make(map[RoutingTableLocalSyncer]struct{}, 0)
+	s.active = config.GetBool("ears.synchronization.active")
+	if !s.active {
+		logger.Info().Msg("Redis Delta Syncer Not Activated")
+	} else {
+		s.client = redis.NewClient(&redis.Options{
+			Addr:     s.redisEndpoint,
+			Password: "",
+			DB:       0,
+		})
+		//s.ListenForSyncRequests()
+		//logger.Info().Msg("Redis Syncer Started")
+	}
+	return s
+}
+
+func (s *RedisDeltaSyncer) RegisterLocalTableSyncer(localTableSyncer RoutingTableLocalSyncer) {
+	// the redis implementation should only ever register itself
+	s.Lock()
+	defer s.Unlock()
+	s.localTableSyncers[localTableSyncer] = struct{}{}
+}
+
+func (s *RedisDeltaSyncer) UnregisterLocalTableSyncer(localTableSyncer RoutingTableLocalSyncer) {
+	s.Lock()
+	defer s.Unlock()
+	delete(s.localTableSyncers, localTableSyncer)
+}
+
+// PublishSyncRequest asks others to sync their routing tables
+
+func (s *RedisDeltaSyncer) PublishSyncRequest(ctx context.Context, routeId string, instanceId string, add bool) {
+	if !s.active {
+		return
+	}
+	cmd := ""
+	if add {
+		cmd = EARS_ADD_ROUTE_CMD
+	} else {
+		cmd = EARS_REMOVE_ROUTE_CMD
+	}
+	sid := uuid.New().String() // session id
+	numSubscribers := s.GetInstanceCount(ctx)
+	// outer go func is so that PublishSyncRequest returns immediately despite the 10 ms wait below
+	// this primarily cause issues when multi route unit tests share the same debug receiver
+	// in practice this may not be an issue
+	go func() {
+		var wg sync.WaitGroup
+		wg.Add(1)
+		// listen for ACKs first ...
+		go func() {
+			if numSubscribers <= 1 {
+				s.logger.Info().Str("op", "PublishSyncRequest").Msg("no subscribers but me - no need to wait for ack")
```

Comment:
Coding style comment: you might want to follow early return pattern instead of having an else bracket for code readability. There are several level of nesting brackets here making things hard to read

Reply:
In fact, you might just want to check numSubscribers right after you get it at line 101 and early return if it is <= 1

Reply:
ok